### PR TITLE
[ML] Wire in incremental training

### PR DESCRIPTION
**Pull request by tveasey**
This wires in incremental training to `CBoostedTreeImpl`. It is intended to be a quick first draft implementation so we can iron out all the details around passing in parameters, old training data, new training data and the old model. Some details of the implementation will likely change a bit.

### CODE REVIEW

File: lib/maths/CBoostedTreeImpl.cc [hunk 1]
```diff
@@ -1453,31 +1751,38 @@ bool CBoostedTreeImpl::selectNextHyperparameters(const TMeanVarAccumulator& loss
...
+
 void CBoostedTreeImpl::restoreBestHyperparameters() {
     m_Regularization = m_BestHyperparameters.regularization();
     m_DownsampleFactor = m_BestHyperparameters.downsampleFactor();
     m_Eta = m_BestHyperparameters.eta();
     m_EtaGrowthRatePerTree = m_BestHyperparameters.etaGrowthRatePerTree();
     m_MaximumNumberTrees = m_BestHyperparameters.maximumNumberTrees();
     m_FeatureBagFraction = m_BestHyperparameters.featureBagFraction();
+    m_PredictionChangeCost = m_BestHyperparameters.predictionChangeCost();
     LOG_TRACE(<< "loss* = " << m_BestForestTestLoss
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeImpl.cc:1786]: Maybe add the new hyperparameters to the log trace as well?

File: lib/maths/CBoostedTreeImpl.cc [hunk 2]
```diff
@@ -1510,15 +1815,19 @@ std::size_t CBoostedTreeImpl::numberHyperparametersToTune() const {
...
         std::ceil(10.0 * std::sqrt(static_cast<double>(numberRows))));
 }
 
+std::size_t CBoostedTreeImpl::numberTreesToRetrain() const {
+    return m_TreesToRetrain.size();
+}
+
 void CBoostedTreeImpl::recordHyperparameters() {
     m_Instrumentation->hyperparameters().s_Eta = m_Eta;
     m_Instrumentation->hyperparameters().s_ClassAssignmentObjective = m_ClassAssignmentObjective;
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeImpl.cc:1833]: I think, the tree topology change regularization hyperparameter should be added here. As well as the prediction change cost hyperparameter

File: lib/maths/CBoostedTreeImpl.cc [hunk 3]
```diff
@@ -1628,6 +1955,22 @@ void CBoostedTreeImpl::skipProgressMonitoringFinalTrain() {
...
+    LOG_TRACE(<< "main loop total number steps = " << totalNumberSteps);
+    m_TrainingProgress = core::CLoopProgress{
+        totalNumberSteps, m_Instrumentation->progressCallback(), 1.0, 1024};
+
+    // Make sure progress starts where it left off.
+    m_TrainingProgress.increment(m_CurrentRound * this->numberTreesToRetrain() * m_NumberFolds);
+}
+
 namespace {
 const std::string VERSION_7_11_TAG{"7.11"};
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeImpl.cc:1975]: Should we add 8.0 version tag here? Or a todo to adjust to whatever version will be current when we merge to master? Otherwise we will probably forget about it by the time.

File: lib/maths/unittest/CBoostedTreeTest.cc [hunk 4]
```diff
@@ -4,6 +4,7 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
+#include "core/CPackedBitVector.h"
```
Comment (valeriy42) [inline lib/maths/unittest/CBoostedTreeTest.cc:7]:
```suggestion
#include <core/CPackedBitVector.h>
```

**Review by valeriy42 — COMMENTED** [review]
Good work on the wiring in incremental training :rocket:  I have some comments mainly concerning the readability of the new code. 

Once all pieces are clear, I think, we should make an effort of splitting non-incremental and incremental code in a subsequent PR. There will be some code duplication, but it significantly simplify the reading flow.

File: include/maths/CBoostedTreeLeafNodeStatistics.h [hunk 5]
```diff
@@ -559,37 +552,35 @@ class MATHS_EXPORT CBoostedTreeLeafNodeStatistics {
         }
 
         //! Get the reduction of the per thread aggregate derivatives.
-        CSplitsDerivatives& reducedDerivatives(std::size_t index) {
+        CSplitsDerivatives& reducedDerivatives() {
```
[the remark this answers is not in the document]
Comment (tveasey) [reply include/maths/CBoostedTreeLeafNodeStatistics.h:555]: At one point I thought I would need to have separate derivatives on old and new training data. So I introduced this in an earlier change on the feature branch. In fact, it turned out that this can all be transparent from this code and handled by the loss functions. So this is just reverting an unnecessary complication (the supplied index was always 0).

File: include/maths/CBoostedTreeImpl.h [hunk 6]
```diff
@@ -259,11 +275,17 @@ class MATHS_EXPORT CBoostedTreeImpl final {
...
-    trainForest(core::CDataFrame& frame,
-                const core::CPackedBitVector& trainingRowMask,
-                const core::CPackedBitVector& testingRowMask,
-                core::CLoopProgress& trainingProgress) const;
+    TNodeVecVecDoubleDoubleVecTr trainForest(core::CDataFrame& frame,
+                                             const core::CPackedBitVector& trainingRowMask,
+                                             const core::CPackedBitVector& testingRowMask,
+                                             core::CLoopProgress& trainingProgress) const;
+
+    //! Retrain the \p treesToRetrain of the best forest.
```
[the remark this answers is not in the document]
Comment (tveasey) [reply include/maths/CBoostedTreeImpl.h:286]: Do you mean `trainForest` and `retrainForest`? If so, they are different. It's probably easiest to just discuss this offline.

File: lib/maths/CBoostedTreeImpl.cc [hunk 7]
```diff
@@ -542,8 +638,27 @@ void CBoostedTreeImpl::initializeTreeShap(const core::CDataFrame& frame) {
     }
 }
 
+void CBoostedTreeImpl::selectTreesToRetrain(const core::CDataFrame& frame) {
+
+    if (m_TreesToRetrain.empty() == false) {
+        return;
+    }
+
+    TDoubleVec probabilities{retrainTreeSelectionProbabilities(
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc:647]: I pondered this as well. Basically, I wanted to find code I could move out of `CBoostedTreeImpl`: it is already rather large. It is also exposed for testing. We could, of course, have a different mechanism for testing internal details of the `CBoostedTreeImpl`. Since this change doesn't actually introduce the function and it'll be a largish change to refactor. I don't want to make it as part of this PR. I'll leave a TODO to investigate whether we move it back into `CBoostedTreeImpl`.

File: include/maths/CBoostedTreeImpl.h [hunk 8]
```diff
@@ -91,16 +94,21 @@ class MATHS_EXPORT CBoostedTreeImpl final {
 
     //! Incrementally train the current model on the values in \p frame.
     //!
-    //! \warning Assumes that train has been called or a trained model has been
-    //! reloaded.
-    bool trainIncremental(core::CDataFrame& frame,
+    //! \warning Assumes that train has already been called or a trained model has
```
[the remark this answers is not in the document]
Comment (tveasey) [reply include/maths/CBoostedTreeImpl.h:98]: See [5260aeb](https://github.com/elastic/ml-cpp/pull/1837/commits/5260aeb2a8e6b0d22c673a9844a1fb13aac145dd).

File: lib/maths/CBoostedTreeImpl.cc (hunk 1 above)
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc:1786]: See [5260aeb](https://github.com/elastic/ml-cpp/pull/1837/commits/5260aeb2a8e6b0d22c673a9844a1fb13aac145dd).

File: lib/maths/CBoostedTreeImpl.cc [hunk 9]
```diff
@@ -1206,16 +1405,36 @@ void CBoostedTreeImpl::candidateRegressorFeatures(const TDoubleVec& probabilitie
...
+void CBoostedTreeImpl::removePredictions(core::CDataFrame& frame,
+                                         const core::CPackedBitVector& trainingRowMask,
+                                         const core::CPackedBitVector& testingRowMask,
+                                         const TNodeVec& tree) const {
+
+    core::CPackedBitVector rowMask{trainingRowMask | testingRowMask};
 
-    using TArgMinLossVec = std::vector<CArgMinLoss>;
+    frame.writeColumns(m_NumberThreads, 0, frame.numberRows(),
+                       [&](TRowItr beginRows, TRowItr endRows) {
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc:1416]: This is the function signature expected for the callback. I agree we should probably be passing by constant reference. The types are lightweight and compared to the work we do in these calls copying relatively small objects here is probably negligible, but it is no more complex to pass by constant reference. Since I'd have to change every single call, and there are quite a few, let's make this change in a separate PR.

File: lib/maths/CBoostedTreeImpl.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc:647]: See [5260aeb](https://github.com/elastic/ml-cpp/pull/1837/commits/5260aeb2a8e6b0d22c673a9844a1fb13aac145dd).

File: lib/maths/CBoostedTreeImpl.cc [hunk 10]
```diff
@@ -734,14 +862,85 @@ CBoostedTreeImpl::trainForest(core::CDataFrame& frame,
...
+    core::CPackedBitVector newTrainingRowMask{trainingRowMask & m_NewTrainingRowMask};
+    auto downsampledRowMask = this->downsample(oldTrainingRowMask) |
+                              this->downsample(newTrainingRowMask);
+    scopeMemoryUsage.add(downsampledRowMask);
+    auto candidateSplits = this->candidateSplits(frame, trainingRowMask);
+    scopeMemoryUsage.add(candidateSplits);
+
+    TDoubleVec losses;
+    losses.reserve(m_TreesToRetrain.size());
+    CTrainForestStoppingCondition stoppingCondition{m_TreesToRetrain.size()};
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc]: Yes good spot. I'd planned to do early stopping, but then decided not to bother (at least initially). See [f77ab66](https://github.com/elastic/ml-cpp/pull/1837/commits/f77ab66e09cc692aa718a4684d0d20c16bf4e602).

File: lib/maths/unittest/CBoostedTreeTest.cc (hunk 4 above)
Comment (tveasey) [reply lib/maths/unittest/CBoostedTreeTest.cc]: See [f77ab66](https://github.com/elastic/ml-cpp/pull/1837/commits/f77ab66e09cc692aa718a4684d0d20c16bf4e602)

File: lib/maths/CBoostedTreeImpl.cc [hunk 11]
```diff
@@ -734,14 +862,85 @@ CBoostedTreeImpl::trainForest(core::CDataFrame& frame,
...
+        retrainedTrees.push_back(std::move(tree));
+        trainingProgress.increment();
+
+        downsampledRowMask = this->downsample(oldTrainingRowMask) |
+                             this->downsample(newTrainingRowMask);
+
+        losses.push_back(this->meanAdjustedLoss(frame, testingRowMask));
+    }
+
+    auto bestLoss = std::min_element(losses.begin(), losses.end()) - losses.begin();
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc]: See [f77ab66](https://github.com/elastic/ml-cpp/pull/1837/commits/f77ab66e09cc692aa718a4684d0d20c16bf4e602)

File: lib/maths/CBoostedTreeImpl.cc [hunk 12]
```diff
@@ -1225,30 +1444,20 @@ void CBoostedTreeImpl::refreshPredictionsAndLossDerivatives(core::CDataFrame& fr
...
-                        double weight{readExampleWeight(row, m_ExtraColumns)};
-                        leafValues_[rootNode.leafIndex(encodedRow, tree)].add(
-                            encodedRow, true /*new example*/, prediction, actual, weight);
-                    }
-                },
-                std::move(leafValues)),
-            &trainingRowMask);
+        TArgMinLossVecVec result(m_NumberThreads, leafValues);
+        if (m_IncrementalTraining) {
+            this->minimumLossLeafValues(frame, trainingRowMask & ~m_NewTrainingRowMask,
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc]: Good suggestion. See [f510256](https://github.com/elastic/ml-cpp/pull/1837/commits/f510256cd275e48e9c4b8af1ae7db720304af453).

File: lib/maths/CBoostedTreeImpl.cc [hunk 13]
```diff
@@ -1262,24 +1471,69 @@ void CBoostedTreeImpl::refreshPredictionsAndLossDerivatives(core::CDataFrame& fr
     LOG_TRACE(<< "tree =\n" << root(tree).print(tree));
 
     core::CPackedBitVector updateRowMask{trainingRowMask | testingRowMask};
+
+    if (m_IncrementalTraining) {
+        this->writeRowDerivatives(frame, updateRowMask & ~m_NewTrainingRowMask,
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc]: Good suggestion. See [f510256](https://github.com/elastic/ml-cpp/pull/1837/commits/f510256cd275e48e9c4b8af1ae7db720304af453).

File: lib/maths/CBoostedTreeImpl.cc (hunk 2 above)
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc:1843]: Also I wasn't actually recording hyperparameters in the loop. Fixed in [4c96cfc](https://github.com/elastic/ml-cpp/pull/1837/commits/4c96cfc59912e06b101a4b3d55d025edf090c385).

File: lib/maths/CBoostedTreeImpl.cc (hunk 3 above)
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc:1991]: Yes. Let's add a TODO. It is probably possible to upgrade state, but we can revisit when we have finalised everything.

Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc:1993]: See [0472b2a](https://github.com/elastic/ml-cpp/pull/1837/commits/0472b2a7a4fa67310d0058f8cde208b0642bda63).

File: include/maths/CBoostedTreeImpl.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (tveasey) [reply include/maths/CBoostedTreeImpl.h]: This was a confusion caused by rather similar function names: I changed `retrainForest` -> `updateForest`. See [e8ea736](https://github.com/elastic/ml-cpp/pull/1837/commits/e8ea7367673aa69fa0d7417c061cb639cde341bc).

File: lib/maths/CBoostedTreeImpl.cc [hunk 14]
```diff
@@ -292,23 +302,102 @@ void CBoostedTreeImpl::train(core::CDataFrame& frame,
...
+        std::tie(lossMoments, std::ignore, numberNodes) = this->crossValidateForest(
+            frame, maximumNumberTrees,
+            [this](core::CDataFrame& frame_, const core::CPackedBitVector& trainingRowMask,
+                   const core::CPackedBitVector& testingRowMask,
+                   core::CLoopProgress& trainingProgress) {
+                return this->retrainForest(frame_, trainingRowMask,
+                                           testingRowMask, trainingProgress);
+            });
+
+        this->captureBestHyperparameters(lossMoments, m_BestForest.size(),
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc:370]: Yes, this was confusing. I improved along these lines in [e8ea736](https://github.com/elastic/ml-cpp/pull/1837/commits/e8ea7367673aa69fa0d7417c061cb639cde341bc).

File: include/maths/CBoostedTreeImpl.h [hunk 15]
```diff
@@ -329,7 +379,11 @@ class MATHS_EXPORT CBoostedTreeImpl final {
     //! Capture the current hyperparameter values.
     void captureBestHyperparameters(const TMeanVarAccumulator& lossMoments,
```
[the remark this answers is not in the document]
Comment (tveasey) [reply include/maths/CBoostedTreeImpl.h:389]: See [e8ea736](https://github.com/elastic/ml-cpp/pull/1837/commits/e8ea7367673aa69fa0d7417c061cb639cde341bc).

File: lib/maths/CBoostedTreeImpl.cc [hunk 16]
```diff
@@ -229,12 +235,16 @@ void CBoostedTreeImpl::train(core::CDataFrame& frame,
             TMeanVarAccumulator lossMoments;
             std::size_t maximumNumberTrees;
             double numberNodes;
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc:237]: I think this naming would be confusing in the context of train from scratch. Hopefully, this is clearer now based on updated function documentation.

File: lib/maths/CBoostedTreeImpl.cc [hunk 17]
```diff
@@ -734,14 +862,85 @@ CBoostedTreeImpl::trainForest(core::CDataFrame& frame,
...
+    scopeMemoryUsage.add(downsampledRowMask);
+    auto candidateSplits = this->candidateSplits(frame, trainingRowMask);
+    scopeMemoryUsage.add(candidateSplits);
+
+    TDoubleVec losses;
+    losses.reserve(m_TreesToRetrain.size());
+    CTrainForestStoppingCondition stoppingCondition{m_TreesToRetrain.size()};
+    TWorkspace workspace{m_Loss->numberParameters()};
+
+    // For feature candidate split set S, for each iteration:
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc]: I reworded slightly. Hopefully, this is clearer now. See [5c0e864](https://github.com/elastic/ml-cpp/pull/1837/commits/5c0e8642f9207e042a477fd4ea8f6f37b03960e8).

**Comment by tveasey** [comment]
Thanks for the review @valeriy42! I've been through all of them and addressed (or in a couple of cases suggested postponing the change to a follow on PR).

> Once all pieces are clear, I think, we should make an effort of splitting non-incremental and incremental code in a subsequent PR.

I agree let's revisit the design down-the-line. I think incremental training is one aspect, but there may be other more effective abstractions we can introduce to split up `CBoostedTreeImpl`, which I think is the real goal. As you say, it'll be easier when we are looking at the final functionality.

**Review by valeriy42 — APPROVED** [review]
LGTM. Good work :+1:
